src/vtrasterizer/TextRenderer.cpp

```diff
@@ -614,89 +554,75 @@ void TextRenderer::renderRasterizedGlyph(crispy::point pen,
+    if (_fontDescriptions.builtinBoxDrawing)
+        return _boxDrawingRenderer.render(position.line, position.column, codepoint, foregroundColor);
 
-    if (attribsChanged || textStartsNewCluster)
-    {
-        if (_textClusterGroup.cellCount)
-            flushTextClusterGroup(); // also increments text start position
-        _textClusterGroup.color = color;
-        _textClusterGroup.style = style;
-    }
-
-    for (char32_t const codepoint: codepoints)
-    {
-        _textClusterGroup.codepoints.emplace_back(codepoint);
-        _textClusterGroup.clusters.emplace_back(_textClusterGroup.cellCount);
-    }
-    _textClusterGroup.cellCount++;
-    if (cellIsEmpty)
-        flushTextClusterGroup(); // also increments text start position
+    return false;
 }
 
-void TextRenderer::flushTextClusterGroup()
+void TextRenderer::renderTextGroup(std::u32string_view codepoints,
+                                   gsl::span<unsigned> clusters,
+                                   vtbackend::CellLocation initialPenPosition,
+                                   TextStyle style,
+                                   vtbackend::RGBColor color)
 {
-    if (!_textClusterGroup.codepoints.empty())
-    {
-        _textRendererEvents.onBeforeRenderingText();
+    if (codepoints.empty())
+        return;
 
-        auto hash = hashTextAndStyle(
-            u32string_view(_textClusterGroup.codepoints.data(), _textClusterGroup.codepoints.size()),
-            _textClusterGroup.style);
-        text::shape_result const& glyphPositions = getOrCreateCachedGlyphPositions(hash);
-        crispy::point pen = _textClusterGroup.initialPenPosition;
-        auto const advanceX = *_gridMetrics.cellSize.width;
+    _textRendererEvents.onBeforeRenderingText();
+    auto _ = crispy::finally { [&]() noexcept {
+        _textRendererEvents.onAfterRenderingText();
+    } };
 
-        for (text::glyph_position const& glyphPosition: glyphPositions)
-        {
-            if (AtlasTileAttributes const* attributes = ensureRasterizedIfDirectMapped(glyphPosition.glyph))
-            {
-                auto const pen1 = applyGlyphPositionToPen(pen, *attributes, glyphPosition);
-                renderRasterizedGlyph(pen1, _textClusterGroup.color, *attributes);
-                pen.x += static_cast<decltype(pen.x)>(advanceX);
-                continue;
-            }
+    auto const hash = hashTextAndStyle(codepoints, style);
+    text::shape_result const& glyphPositions =
+        getOrCreateCachedGlyphPositions(hash, codepoints, clusters, style);
+    crispy::point pen = _gridMetrics.mapBottomLeft(initialPenPosition);
+    auto const advanceX = *_gridMetrics.cellSize.width;
 
-            auto const hash = hashGlyphKeyAndPresentation(glyphPosition.glyph, glyphPosition.presentation);
+    for (text::glyph_position const& glyphPosition: glyphPositions)
+    {
+        if (AtlasTileAttributes const* attributes = ensureRasterizedIfDirectMapped(glyphPosition.glyph))
```

Comment:
we can replace it with 
```suggestion
        if (auto attributes = ensureRasterizedIfDirectMapped(glyphPosition.glyph))
```